### DISCUSSION TITLE
Draggable Point Chart

### DISCUSSION THREAD

**Discussion by natankeddem** [zauberzeug/nicegui#1170] [Ideas / Feature Requests]
I think a useful addition to the graphical controls already in NiceGUI would be a chart with traces that could be dragged. Something similar to this:
[Draggable Point Chart](https://i.stack.imgur.com/3BiM5.gif)

I am not sure which library would be the easiest path forward. I do see there is a HighCharts module which implements this already which might make it a good candidate?
[HighCharts dragDrop Module](https://api.highcharts.com/highcharts/plotOptions.series.dragDrop)

**Comment by falkoschindler** [comment]
Thanks for the suggestion, @natankeddem!
HighCharts might be just the right tool, especially since we're using it anyway for `ui.chart`. It's probably a matter of forwarding the right events to the server.

**Comment by natankeddem** [comment]
Okay so I am going to give updating the chart element with this feature a shot. I am going to document my journey here because I figure that others might find it useful as reference. First thing first; let's get something going in just Highcharts. I made a simple example of a line chart that you can drag the points and the console will print the interaction using the plotOptions.series.point.events.drag event:
[https://jsfiddle.net/t9d38yke/22/](https://jsfiddle.net/t9d38yke/22/)

Next step, let's send those options to the NiceGUI chart element and see what happens. Making sure to import draggable-points.js using "extras=["draggable-points"]".

```py
from nicegui import ui
from random import random

drag = """function (e) {
												console.log((this.name || this.id) + 
                        '=' + ' x/x2/y: ' + e.newPoint.x + '/' + e.newPoint.x2 + 
                        '/' + e.newPoint.y)
                    }"""

chart = ui.chart(
    {
        "chart": {"animation": "false"},
        "xAxis": {"categories": ["1", "2", "3", "4", "5"]},
        "yAxis": {"softMin": -200, "softMax": 400},
        "plotOptions": {
            "series": {
                "stickyTracking": "false",
                "dragDrop": {"draggableX": "true", "draggableY": "true"},
                "point": {"events": {"drag": drag}},
            }
        },
        "series": [{"data": [0, 71.5, 106.4, 129.2, 144.0]}],
    },
    extras=["draggable-points"],
).classes("w-full")


def update():
    chart.options["series"][0]["data"].append(random())
    chart.update()


ui.button("Update", on_click=update)

ui.run(dark=True)
```
Graph renders and is draggable, good so far. Now to the issues. My console print no longer works, so apparently that isn't getting processed. I will need to spend some more time figuring out things here.

Ideally I think we want an on_change that gets activate with the callback:
https://api.highcharts.com/class-reference/Highcharts#.PointDropCallbackFunction
this would then trigger an update of the `Chart.options["series"]`(py) with data from `this.chart.series`(js). I haven't quite figured out how data travels from `PointDropCallbackFunction->this.chart.series(js)->Chart.options["series"](py)` or even how to get `PointDropCallbackFunction` to trigger.

**Comment by natankeddem** [comment]
Getting closer now. I am now injecting the drop callback into options before instantiating the Highcharts object in charts.js:
```js
export default {
  template: "<div></div>",
  mounted() {
    setTimeout(() => {
      const imports = this.extras.map((extra) => import(window.path_prefix + extra));
      Promise.allSettled(imports).then(() => {
        this.seriesCount = this.options.series ? this.options.series.length : 0;
        this.options.plotOptions.series.point = {
          events: {
            drop: function (e) {
              console.log("hi");
              this.$emit("change", e.target.series.chart.series); // this doesn't work how do I emit here?
            },
          },
        };
        this.chart = Highcharts[this.type](this.$el, this.options);
        this.chart.reflow();
      });
    }, 0); // NOTE: wait for window.path_prefix to be set in app.mounted()
  },
  beforeDestroy() {
    this.destroyChart();
  },
  beforeUnmount() {
    this.destroyChart();
  },
  methods: {
    update_chart() {
      if (this.chart) {
        while (this.seriesCount > this.options.series.length) {
          this.chart.series[0].remove();
          this.seriesCount--;
        }
        while (this.seriesCount < this.options.series.length) {
          this.chart.addSeries({}, false);
          this.seriesCount++;
        }
        this.chart.update(this.options);
      }
    },
    destroyChart() {
      if (this.chart) {
        this.chart.destroy();
      }
    },
  },
  props: {
    type: String,
    options: Object,
    extras: Array,
  },
};
```
I altered the charts.py to receive the emit from the backend and map it an on_change method:
```py
    def __init__(
        self, options: Dict, *, type: str = "chart", extras: List[str] = [], on_change: Optional[Callable] = None
    ) -> None:
        """Chart

        An element to create a chart using `Highcharts <https://www.highcharts.com/>`_.
        Updates can be pushed to the chart by changing the `options` property.
        After data has changed, call the `update` method to refresh the chart.

        By default, a `Highcharts.chart` is created.
        To use, e.g., `Highcharts.stockChart` instead, set the `type` property to "stockChart".

        :param options: dictionary of Highcharts options
        :param type: chart type (e.g. "chart", "stockChart", "mapChart", ...; default: "chart")
        :param extras: list of extra dependencies to include (e.g. "annotations", "arc-diagram", "solid-gauge", ...)
        """
        super().__init__()
        self._props["type"] = type
        self._props["options"] = options
        self._props["extras"] = extras
        self.libraries.extend(library for library in self.extra_libraries if library.path.stem in extras)
        self.on("change", on_change)
```
I should now theoretically be able to call a python lambda with on_change:
```py
from nicegui import ui

chart = ui.chart(
    {
        "chart": {"animation": "false"},
        "xAxis": {"categories": ["1", "2", "3", "4", "5"]},
        "yAxis": {"softMin": -200, "softMax": 400},
        "plotOptions": {
            "series": {"stickyTracking": "false", "dragDrop": {"draggableX": "true", "draggableY": "true"}}
        },
        "series": [{"data": [0, 71.5, 106.4, 129.2, 144.0]}],
    },
    extras=["draggable-points"],
    on_change=lambda e: ui.notify(f"The value changed to {e}."),
)

ui.run(dark=True)

```

My drop callback is firing and I am getting console messages from it. Slow but steady progress. So my main issue now is my rusty/bad/nonexistent javascript/vue knowledge. I am not sure how to emit from the drop callback. I think it has something to do with the fact the Highcharts isn't really a vue object which means its callback has no `this` scope. I will need to spend a bit more time figuring this out. This looks like my issue but I need to digest a bit more:
[https://github.com/highcharts/highcharts-vue/issues/62](https://github.com/highcharts/highcharts-vue/issues/62)

**Comment by natankeddem** [comment]
Well I probably should have tried a bit harder before typing the last message up. But I like this stream of consciousness thing I have going here. Just a few minutes later I found the `=>` operator. Now using this we have the following chart.js which seems to work:
```py
export default {
  template: "<div></div>",
  mounted() {
    setTimeout(() => {
      const imports = this.extras.map((extra) => import(window.path_prefix + extra));
      Promise.allSettled(imports).then(() => {
        this.seriesCount = this.options.series ? this.options.series.length : 0;
        this.options.plotOptions.series.point = {
          events: {
            drop: () => {
              console.log("hi");
              this.$emit("change", this.chart.options.series[0].data);
            },
          },
        };
        this.chart = Highcharts[this.type](this.$el, this.options);
        this.chart.reflow();
      });
    }, 0); // NOTE: wait for window.path_prefix to be set in app.mounted()
  },
  beforeDestroy() {
    this.destroyChart();
  },
  beforeUnmount() {
    this.destroyChart();
  },
  methods: {
    update_chart() {
      if (this.chart) {
        while (this.seriesCount > this.options.series.length) {
          this.chart.series[0].remove();
          this.seriesCount--;
        }
        while (this.seriesCount < this.options.series.length) {
          this.chart.addSeries({}, false);
          this.seriesCount++;
        }
        this.chart.update(this.options);
      }
    },
    destroyChart() {
      if (this.chart) {
        this.chart.destroy();
      }
    },
  },
  props: {
    type: String,
    options: Object,
    extras: Array,
  },
};

```
Match that up to the previous py files/changes and now we have a working proof of concept. Where do you guys want to go from here? Should we extend the current chart element out with this feature? Should this be a new element all together?

**Reply by natankeddem** [reply]
In addition, what should the on_change return? We can have:
- Information on just the single point that has changed.
- Information on just the series that has changed.
- A full report of all series.
- Some mix of the previous options.
I like just returning the specific series that has changed as a good compromise between the more extreme options. 

We could also start obviouscating some value data/series access to this element if we wanted. Instead of having the user pass a full options dictionary with configuration and series data. We could break these down into 2 attributes and have the user interact with one or the other independently which might make series updating more similar to value interactions in other elements like ui.number.
